src/ext/language_tools.js

```diff
@@ -141,19 +141,35 @@ var doLiveAutocomplete = function(e) {
         if (hasCompleter && !util.getCompletionPrefix(editor))
             editor.completer.detach();
     }
-    else if (e.command.name === "insertstring") {
-        var prefix = util.getCompletionPrefix(editor);
-        // Only autocomplete if there's a prefix that can be matched or previous char is trigger character 
-        var triggerAutocomplete = util.triggerAutocomplete(editor);
-        if ((prefix || triggerAutocomplete) && !hasCompleter) {
-            var completer = Autocomplete.for(editor);
-            // Set a flag for auto shown
-            completer.autoShown = true;
-            completer.showPopup(editor);
+    else if (e.command.name === "insertstring" && !hasCompleter) {
+        lastExecEvent = e;
+        var delay = e.editor.$liveAutocompletionDelay;
+        if (delay) {
+            liveAutocompleteTimer.delay(delay);
```

Comment:
This logic is only for the initial case when we show the popup, right? After it, `Autocomplete` logic handles the calls to the completion providers. But those calls won't be delayed anymore, right? (e.g. the `changeTimer` there hard codes 0 as the delay). We also want to debounce them with the same delay, as far as I understand. Seems like we would be better with the delaying logic implemented fully on the `Autocomplete` side.

Reply:
Autocomplete doesn't call completion providers again, and filters the results provided after the first call.

Reply:
True, didn't consider it. But it seems like this logic might actually be undesirable for some use cases - when using ace to support some LSP servers we might need to call completers again even for the already opened popup.